crates/codegen/schema/src/types/precedence_parser.rs

```diff
@@ -0,0 +1,51 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::parser::ParserDefinition;
+
+#[allow(dead_code)]
+pub type PrecedenceParserRef = std::rc::Rc<PrecedenceParser>;
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields)]
+pub struct PrecedenceParser {
+    #[serde(default)]
+    pub name: Option<String>,
+    #[serde(flatten)]
+    pub definition: PrecedenceParserDefinition,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct PrecedenceParserDefinition {
+    #[schemars(title = "Operator Definitions")]
+    pub operators: Vec<OperatorDefinition>,
+
+    #[schemars(title = "Primary Expression References")]
+    pub primary_expressions: Vec<Reference>,
```

Comment:
Similarly for `primary_expressions`. Is there a reason for the extra type? vs `primary_expressions: Vec<String>` directly.

Reply:
Thinking about it, I like that the yaml uses `reference: XXX` syntax here, to indicate it is actually a reference, similar to all other grammar expressions that use `ParserDefinition::Reference` directly.

But I wonder if there is a technical limitation/intentional design to only include references here (as a special/new type), rather than just doing `primary_expressions: Vec<ParserRef>`. Not only this will eliminate the additional type, but it will use the same syntax as the rest of the grammar. Easier for the user, and simpler code for validation/visitors.

Reply:
Is `primary_expressions` array intended to be alternatives (`choice:`)? What is the use case?

I see the grammar still defines `PrimaryExpression` as a single `reference:` to a `choice:`, instead of inlining them in the parent operator definition, which is great, as it doesn't duplicate its contents between multiple versions.

If not, and it is intended to be generic, and leave that to the user, I wonder if something like `primary_expressions: ParserRef` would be better in this case:

- No new types, so simpler visitors and validation code, rather than the new `Reference`.
- Easier on the user to remember and write.
- Can easily contain a `choice:` or a `reference:` to a top-level defined one, like what we have here.

Reply:
Fixing in #368